dataproc/README.md

```diff
@@ -1,20 +1,94 @@
+j = dataproc.hail_dataproc_job(
+    batch=b,
+    script=script,
+    packages=['cpg_workflows', 'google', 'fsspec', 'gcloud'],
+    num_workers=2,
+    num_secondary_workers=20,
+    job_name=job_name,
+    depends_on=depends_on,
+    scopes=['cloud-platform'],
+    pyfiles=pyfiles,
+)
+```
+
+## Docker image
+
+The dataproc image is effectively a driver image for the _setting up_, _submitting to_, and _spinning down_ the cluster. It doesn't do anything too fancy, but it's useful to keep this disconnected from the regular analysis-runner driver image to avoid any arbitrary changes to pip_dependencies in the deploy_config (`/usr/local/lib/python3.10/dist-packages/hailtop/hailctl/deploy.yaml`).
+
+Note that our [Hail fork](https://github.com/populationgenomics/hail) at `HEAD` is used to build the image. You can use the `COMMIT_HASH` arg to build an image with a specific version of hail instead of the latest version, which is what `HEAD` will use.
+
+We manually build our version of Hail in the dataproc container. (Ideally it would be good to use a multistage build process to reduce this image size, but we don't want to support dataproc into the indefinite future).
+
+## Spinning up a dataproc cluster
+
+A dataproc cluster is spun up within a specific dataset's GCP project for billing reasons.
+
+We call `hailctl dataproc start`, as configured in the `analysis_runner/dataproc` module. We specify a number of default packages in this module as a sensible default. We by default specify the init script (`gs://cpg-common-main/hail_dataproc/${HAIL_VERSION}/`), but you can override this on cluster configuration.
+
+### Initialization script
+
+The scripts in [`init_scripts`](init_scripts) are used to install dependencies on Dataproc master nodes through the `init` parameter of the `setup_dataproc` function. The scripts get copied to `gs://cpg-common-main/hail_dataproc/${HAIL_VERSION}` [automatically](../.github/workflows/copy_dataproc_init_scripts.yaml) on pushes to `main`.
+
+## Updating dataproc
+
+When you're trying to update the default version of dataproc, you should:
+
+1. Bump the `DEFAULT_HAIL_VERSION` in `analysis_runner/dataproc.py`
+    * Side note: hail must be released before this happens, including the wheel at `gs://cpg-hail-ci/wheels/hail-{HAIL_VERSION}-py3-none-any.whl`
+2. Completely release the analysis-runner CLI (merge to main with a bumpversion commit)
+3. Rebuild the dataproc image, using the [command below](#rebuilding-the-dataproc-driver-image)
+4. Rebuild the analysis-runner driver image
+
+Note, we support specifying the hail_version, but only a very select number of versions are available (due to the init scripts not always being updated).
+
+### Rebuilding the dataproc driver image
 
 ```sh
 gcloud config set project analysis-runner
-HAIL_VERSION=0.2.110
+# grab the HAIL_VERSION from analysis_runner/dataproc.py
+HAIL_VERSION=$(grep "HAIL_VERSION = '" analysis_runner/dataproc.py | awk -F\' '{print $2}')
```

Comment:
It still works as is, but maybe the same pattern as the other one?

```suggestion
HAIL_VERSION=$(grep "DEFAULT_HAIL_VERSION = '" analysis_runner/dataproc.py | awk -F\' '{print $2}')
```

FYI for next time: I tend to do stuff like this in one via `sed`. E.g. although dropping the quotes makes this messier:

```
sed -n "/^DEFAULT_HAIL_VERSION/s/^[^']*//p" analysis_runner/dataproc.py
```

Reply:
I asked copilot to write this for me 🙃